platform/mellanox/mlnx-platform-api/sonic_platform/component.py

```diff
@@ -110,9 +151,84 @@ def get_firmware_version(self):
         return result
 
 
+    def _check_onie_version(self):
+        # check ONIE version. To update ONIE requires version 5.2.0016 or later.
+        machine_conf = self._read_generic_file(MACHINE_CONF_FILE, MACHINE_CONF_MAX_SIZE)
+
+        try:
+            machine_info = get_machine_info()
+            onie_version_string = machine_info['onie_version']
+            m = re.search(self.ONIE_VERSION_PARSE_PATTERN, onie_version_string)
+            onie_major = m.group(self.ONIE_VERSION_MAJOR_OFFSET)
+            onie_minor = m.group(self.ONIE_VERSION_MINOR_OFFSET)
+            onie_release = m.group(self.ONIE_VERSION_RELEASE_OFFSET)
+        except AttributeError as e:
+            print("Failed to parse ONIE version by {} from {} due to {}".format(
+                                self.ONIE_VERSION_PARSE_PATTERN, machine_conf, repr(e)))
+            return False
+        if onie_major < self.ONIE_REQUIRED_MAJOR or onie_minor < self.ONIE_REQUIRED_MINOR or onie_release < self.ONIE_REQUIRED_RELEASE:
+            print("ONIE {}.{}.{} or later is required".format(self.ONIE_REQUIRED_MAJOR, self.ONIE_REQUIRED_MINOR, self.ONIE_REQUIRED_RELEASE))
+            return False
+
+        return True
+
+
+    def install_firmware(self, image_path):
+        """
+        Installs firmware to the component
+
+        Args:
+            image_path: A string, path to firmware image
+
+        Returns:
+            A boolean, True if install was successful, False if not
+        """
+        # check ONIE version requirement
+        if not self._check_onie_version():
+            return False
+
+        # check whether the file exists
+        if not self._check_file_validity(image_path):
+            return False
+
+        # do the real work.
+        try:
+            # check whether there has already been some images pending
+            # if yes, remove them
+            result = self._get_command_result(ONIE_FW_UPDATE_CMD_SHOW)
+            pending_list = result.split("\n")
+            for pending in pending_list:
+                m = re.match(self.BIOS_PENDING_UPDATE_PATTERN, pending)
+                if m is not None:
+                    pending_image = m.group(1)
+                    self._get_command_result(ONIE_FW_UPDATE_CMD_REMOVE.format(pending_image))
+                    print("Image {} which is already pending to upgrade has been removed".format(pending_image))
+
+            result = subprocess.call(ONIE_FW_UPDATE_CMD_ADD.format(image_path).split())
+            if result:
+                return False
+            result = subprocess.call(ONIE_FW_UPDATE_CMD_UPDATE.split())
```

Comment:
@stephenxs why not to use _get_command_result here?

Reply:
we need to print the output to screen.  get command result returns output for parsing or comparing. they're different. 
all other cases fall into this.

Reply:
BTW, I'm not going to change the way we capture output when upgrading CPLD. current way works and I don't have more time on it :(

Reply:
@stephenxs ok. Let's leave it as is.